csrc/scheduler/reduction.cpp

```diff
@@ -507,40 +507,253 @@ std::shared_ptr<ReductionParams> innerReductionHeuristic(
+
+  if (rparams->multiple_reds_per_blk) {
+    rparams->block_dim_iter_dom = ParallelType::TIDx;
+  }
+
+  rparams->grid_dim_iter_dom =
+      flip_grid ? ParallelType::BIDy : ParallelType::BIDx;
+  if (hp.gidim > (flip_grid ? scheduler_utils::y_grid_limit
+                            : scheduler_utils::x_grid_limit)) {
+    rparams->split_grid_dim_iter_dom_outer = true;
+    if (flip_grid) {
+      gdimy = scheduler_utils::y_grid_limit;
+    } else {
+      gdimx = scheduler_utils::x_grid_limit;
+    }
+  }
+
+  rparams->flip_grid = flip_grid;
+
+  if (rparams->cross_block_inner_reduction) {
+    if (rparams->block_dim_iter_dom == ParallelType::TIDx) {
+      rparams->block_dim_inner_reduction = ParallelType::TIDy;
+    } else {
+      rparams->block_dim_inner_reduction = ParallelType::TIDx;
+    }
+  }
+
+  rparams->unroll_factor_inner_reduction = hp.redu_unroll_factor;
+
+  rparams->unroll_factor_iter_dom = hp.iter_unroll_factor;
+  rparams->vectorize_iter_dom = hp.iter_unroll_factor > 1;
+
+  rparams->lparams = LaunchParams(
+      gdimx,
+      gdimy,
+      LaunchParams::UNINITIALIZED_VAL,
+      rparams->multiple_reds_per_blk ? hp.bdimx : hp.bdimy,
+      rparams->multiple_reds_per_blk ? hp.bdimy
+                                     : LaunchParams::UNINITIALIZED_VAL,
+      LaunchParams::UNINITIALIZED_VAL);
+
+  if (isDebugDumpEnabled(DebugDumpOption::SchedulerDebug)) {
+    debug() << hp.toString() << std::endl;
+    debug() << rparams->toString() << std::endl;
+  }
+  return rparams;
+}
+
+OuterReduHeuristicParas getBlockOuterReduction(
+    int64_t vectorize_factor,
+    int64_t sm_count,
+    int64_t max_unroll,
+    int64_t max_threads_per_block,
+    int64_t total_iteration_numel,
+    int64_t total_reduction_numel) {
+  OuterReduHeuristicParas hp(total_iteration_numel, total_reduction_numel);
+
+  int64_t sm_count_pow2 = scheduler_utils::lastPow2(sm_count);
+  // Step-1, set iteration dim
+  // (1) start with bdimx = 8, gidim = 32, iter_unroll = 1.
+  // starts bdimx from 8, ensures each warp spans at most into 4 different rows.
+  // when each thread reads 16 bytes, each warp does 4 transactions each
+  // with 128 bytes. This is the maximum global memory transaction size
+  // for each warp.
+  // starts gidim from 32 and iter_unroll from 1, defers fully vectorization
+  // to SM usage is high enough.
+  hp.bdimx = std::min(8L, hp.iDimAvail());
+  hp.gidim = std::min(std::min(32L, sm_count_pow2), hp.iDimAvail());
+  hp.iter_unroll_factor = 1;
+
+  // (2) increase iter_unroll to its maximum following two rules:
+  // (2.1) ensure divisible split
+  // (2.2) leave enough blocks to saturate the device.
+  // For example on any GPU with more than 32 SMs
+  // bdimx-vect-gidim = 8-1-32  = 256
+  // bdimx-vect-gidim = 8-2-32  = 512
+  // bdimx-vect-gidim = 8-2-64  = 1024
+  // bdimx-vect-gidim = 8-4-64  = 2048
+  // bdimx-vect-gidim = 8-4-128 = 4096
+  // bdimx-vect-gidim = 8-8-128 = 8192
+  int64_t max_iter_unroll = vectorize_factor;
```

Comment:
nitpick, we can just use vectorize_factor below. Is this renaming trying to make it easier to read? maybe consider renaming the argument of the function.

Reply:
just to keep consistent with other schedulers, will revise when refactor/clean up reduction para names.